packages/search-ui/src/SearchDriver.js

```diff
@@ -219,85 +222,147 @@ export default class SearchDriver {
+
+    // We bail on making state updates if "isLoading" is true, which implies
+    // that there is an active API request that has been made and is pending.
+    //
+    // The net effect of this is that the user interaction effectively becomes
+    // "locked" while API requests are in flight, to avoid any odd race
+    // conditions.
+    //
+    // NOTE: This may be unnecessary now that we have debounce
+    // "_makeSearchRequest".
     if (isLoading && !ignoreIsLoadingCheck) return;
 
     this._setState({
       current,
       error: "",
       filters,
-      isLoading: true,
       resultsPerPage,
       searchTerm,
       sortDirection,
       sortField
     });
 
-    const requestId = this.requestSequencer.next();
-
-    const queryConfig = {
-      ...this.searchQuery,
-      facets: removeConditionalFacets(
-        this.searchQuery.facets,
-        this.searchQuery.conditionalFacets,
-        filters
-      )
-    };
-
-    const requestState = filterSearchParameters(this.state);
-
-    return this.events.search(requestState, queryConfig).then(
-      resultState => {
-        if (this.requestSequencer.isOldRequest(requestId)) return;
-        this.requestSequencer.completed(requestId);
+    this._makeSearchRequest({
+      current,
+      filters,
+      resultsPerPage,
+      searchTerm,
+      skipPushToUrl,
+      sortDirection,
+      sortField
+    });
+  };
 
-        // Results paging start & end
-        const { totalResults } = resultState;
-        const start =
-          totalResults === 0 ? 0 : (current - 1) * resultsPerPage + 1;
-        const end =
-          totalResults <= start + resultsPerPage
-            ? totalResults
-            : start + resultsPerPage - 1;
-
-        this._setState({
-          isLoading: false,
-          resultSearchTerm: searchTerm,
-          pagingStart: start,
-          pagingEnd: end,
-          ...resultState,
-          wasSearched: true
-        });
-
-        if (this.hasA11yNotifications) {
-          const messageArgs = { start, end, totalResults, searchTerm };
-          this.actions.a11yNotify("searchResults", messageArgs);
-        }
+  /**
+   * This method is separated out from _updateSearchResults so that it
+   * can be debounced.
+   *
+   * By debouncing our API calls, we can effectively allow action chaining.
+   *
+   * For Ex:
+   *
+   * If a user needs to make multiple filter updates at once, they could
+   * do so by calling an action 3 times in a row:
+   *
+   *   addFilter("states", "California");
+   *   addFilter("states", "Nebraska");
+   *   addFilter("states", "Pennsylvania");
+   *
+   * We don't want to make 3 separate API calls like that in quick succession,
+   * so we debounce the API calls.
+   *
+   * Application state updates are performed in _updateSearchResults, but we
+   * wait to make the actual API calls until all actions have been called.
+   */
+  _makeSearchRequest = DebounceManager.debounce(
+    0,
```

Comment:
Are we concerned at all about this `0` wait time? Do we want to give it any breathing room, e.g. 3-10, or maybe make it customizable/passable by the developer?

If I remember correctly, 0 just kicks it to the end of the event loop/queue - so if that's all we need, no worries 👍

Reply:
That was my intent, just kicking it to the end of the loop so that you can call multiple actions in a row. I don't actually want to delay processing these by any length of time. Does that make sense?

Reply:
Gotcha. The only even possible edge case scenario I can think of is a Starcraft player with very high APM clicking 2-3 filters in a row within 5ms of each other, and a developer maybe wanting to batch for that, but that's such a ridiculous edge case you should just ignore me haha.